### PR TITLE
Allow building for the web

### PR DESCRIPTION
**Pull request by Mubelotix**

### CODE REVIEW

File: webpki-ccadb/Cargo.toml [hunk 1]
```diff
@@ -20,3 +20,7 @@ serde = { version = "1.0.183", features = ["derive"] }
 webpki = { workspace = true }
 x509-parser = { workspace = true }
 yasna = { workspace = true }
+
+[features]
+default = []
+web = ["pki-types/web"]
```
Comment (cpu) [inline webpki-ccadb/Cargo.toml:26]: Do you have a usecase for building this crate for a web target? It's primarily used for codegen for the data vendored into the webpki-roots crate.

File: webpki-roots/Cargo.toml [hunk 2]
```diff
@@ -22,3 +22,7 @@ webpki = { workspace = true }
 webpki-ccadb = { path = "../webpki-ccadb" }
 x509-parser = { workspace = true }
 yasna = { workspace = true }
+
+[features]
+default = []
+web = ["pki-types/web"]
```
Comment (cpu) [inline webpki-roots/Cargo.toml:28]: I think CI would benefit from an update that adds `--all-features` coverage.

**Review by cpu — COMMENTED** [review]
:wave: Thanks for the PR. I had a couple initial comments.

File: webpki-roots/Cargo.toml (hunk 2 above)
Comment (djc) [reply webpki-roots/Cargo.toml:28]: I see no point in defining an empty `default` feature.

**Comment by ctz** [comment]
I don't think this is necessary? There's no reason that a crate which takes a dependency must re-export or otherwise control the dependency's features.

If you want to set `pki-types/web` in your application, just write in your Cargo.toml: `pki-types { version = "*", features = ["web"] }` -- job done, no need for any other crates to change.

(The `version = "*"` is only appropriate if you do not use any of the APIs from `pki-types` yourself.)

**Comment by Mubelotix** [comment]
> I don't think this is necessary? There's no reason that a crate which takes a dependency must re-export or otherwise control the dependency's features.
> 
> If you want to set `pki-types/web` in your application, just write in your Cargo.toml: `pki-types = { version = "*", features = ["web"] }` -- job done, no need for any other crates to change.
> 
> (The `version = "*"` is only appropriate if you do not use any of the APIs from `pki-types` yourself.)

This is what I have been doing before but it doesn't feel right having to dig into the dependency tree to fix all compile errors by adding features with no standardized name

**Comment by Mubelotix** [comment]
I made a similar PR on [rustls#1921](https://github.com/rustls/rustls/pull/1921) btw

**Comment by djc** [comment]
> I don't think this is necessary? There's no reason that a crate which takes a dependency must re-export or otherwise control the dependency's features.

I agree it's not strictly necessary, but it usually makes things easier for the consumer at little cost to the maintainer?

**Review by djc — APPROVED** [review]
I think this would be okay to have.

What we really want here is target-dependent features, which I think we can't do? I don't remember why we didn't go with that in the rustls-pki-types [PR](https://github.com/rustls/pki-types/pull/32)?

**Comment by ctz** [comment]
> I agree it's not strictly necessary, but it usually makes things easier for the consumer at little cost to the maintainer?

I think a basic level of testing is that all combinations of a crate's features are tested.  If we have features which do not vary the code or dependencies, that adds new dimensions to that testing for no benefit and eventually makes that style of testing implausible.

I think a crate should only have features that either change its dependencies or behaviour; but not just pass through to some other crate's features. I especially think that webpki-roots is an odd choice to do that.

**Comment by djc** [comment]
> I think a crate should only have features that either change its dependencies or behaviour; but not just pass through to some other crate's features. I especially think that webpki-roots is an odd choice to do that.

That is fair. And if we add a flag for this in rustls, we probably wouldn't strictly need it here?

**Comment by djc** [comment]
Agreed, let's close this for now.

**Pull request closed without merging**